evm/src/executor/abi.rs

```diff
@@ -680,6 +681,341 @@ pub static HARDHAT_CONSOLE_SELECTOR_PATCHES: Lazy<HashMap<Selector, Selector>> =
+                }
+            }
+            continue
+        }
+        if expect_fmt && (*c == b's' || *c == b'd' || *c == b'i' || *c == b'o') {
+            let fspec = match *c {
+                b's' => FormatSpec::String,
+                b'd' => FormatSpec::Number,
+                b'i' => FormatSpec::Integer,
+                b'o' => FormatSpec::Object,
+                _ => unreachable!(),
+            };
+            let v = f(fspec);
+            result.push_str(&v);
+            return (result, pos + 1)
+        } else {
+            result.push(*c as char);
+        }
+    }
+    write!(result, " {default}").unwrap();
+    (result, spec.len())
+}
+
+trait ConsoleLogFragment {
+    fn log_fragment(&self, spec: &str) -> (String, usize);
+}
+
+impl ConsoleLogFragment for String {
+    fn log_fragment(&self, spec: &str) -> (String, usize) {
+        let formatter = |spec| match spec {
+            FormatSpec::String => self.clone(),
+            FormatSpec::Object => format!("'{}'", self.clone()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        };
+        format_fragment(spec, formatter, self)
+    }
+}
+
+impl ConsoleLogFragment for bool {
+    fn log_fragment(&self, spec: &str) -> (String, usize) {
+        let formatter = |spec| match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number => (*self as i32).to_string(),
+            FormatSpec::Integer => String::from("NaN"),
+        };
+        format_fragment(spec, formatter, &self.pretty())
+    }
+}
+
+impl ConsoleLogFragment for U256 {
+    fn log_fragment(&self, spec: &str) -> (String, usize) {
+        let formatter = |_spec| self.pretty();
+        format_fragment(spec, formatter, &self.pretty())
+    }
+}
+
+impl ConsoleLogFragment for I256 {
+    fn log_fragment(&self, spec: &str) -> (String, usize) {
+        let formatter = |_spec| self.pretty();
+        format_fragment(spec, formatter, &self.pretty())
+    }
+}
+
+impl ConsoleLogFragment for Address {
+    fn log_fragment(&self, spec: &str) -> (String, usize) {
+        let formatter = |spec| match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        };
+        format_fragment(spec, formatter, &self.pretty())
+    }
+}
+
+impl ConsoleLogFragment for Bytes {
+    fn log_fragment(&self, spec: &str) -> (String, usize) {
+        let formatter = |spec| match spec {
+            FormatSpec::String => self.pretty(),
+            FormatSpec::Object => format!("'{}'", self.pretty()),
+            FormatSpec::Number | FormatSpec::Integer => String::from("NaN"),
+        };
+        format_fragment(spec, formatter, &self.pretty())
+    }
+}
+
+pub trait HardhatConsoleLogf {
+    fn logf(&self) -> String;
```

Comment:
I was trying to avoid Boxing each `UiFmt` field in the Log in order to be passed as a `dyn UiFmt`. Though now I see that's not even necessary. Will simplify.